src/colvarproxy.cpp

```diff
@@ -597,7 +597,20 @@ colvarproxy::colvarproxy()
 }
 
 
-colvarproxy::~colvarproxy() {}
+colvarproxy::~colvarproxy() {
+  if (smp_enabled() == COLVARS_OK && smp_thread_id() > 0) {
+    // Nothing to do on non-master threads
+    return;
+  }
+  std::list<std::string>::iterator    osni = output_stream_names.begin();
+  std::list<std::ostream *>::iterator osi  = output_files.begin();
+  for ( ; osi != output_files.end(); osi++, osni++) {
+    ((std::ofstream *) (*osi))->close();
+    delete *osi;
+  }
+  output_files.clear();
+  output_stream_names.clear();
+}
```

Comment:
Yeah, we just need to make sure the destructor is actually called at some point. I have a NAMD patch for that.

Reply:
Would your patch work with multiple `run` commands using the same topology?

Reply:
Yes, I would say it's a pretty safe bet. Here is the patch in its complex glory:
```
--- a/src/ComputeMgr.C
+++ b/src/ComputeMgr.C
@@ -135,6 +135,7 @@ ComputeMgr::ComputeMgr()
 ComputeMgr::~ComputeMgr(void)
 {
     delete computeNonbondedWorkArrays;
+    delete masterServerObject;
 }
```
My reasoning is: if  `ComputeMgr` allocates `masterServerObject` with `new`, it should delete it when destroyed. That seems safe enough.

Reply:
I've checked, `colvarproxy` is destroyed correctly in LAMMPS and Gromacs. The NAMD PR is here: https://charm.cs.illinois.edu/gerrit/c/namd/+/5391
The VMD case is tricky because the proxy is created by a Tcl script command, so the pointer does not belong to a C++ object. There is no "final cleanup" routine for the Tcl scripting side of VMD.

Reply:
I'm not sure that the destructor is always called when needed (i.e. at the end of a `run` command even though it is followed by another command).  In NAMD we have this:
```
  // NAMD does not destruct GlobalMaster objects, so we must remember
  // to write all output files at the end of a run
  if (step == simparams->N) {
    colvars->write_restart_file(cvm::output_prefix()+".colvars.state");
    colvars->write_output_files();
  }
```
And in LAMMPS we have this:
```
void FixColvars::post_run()
{
  if (me == 0) {
    proxy->write_output_files();
  }
}
```
which is a wrapper for the same two calls above.

Even after your change to NAMD (which does make a lot of sense for memory management) I am really not sure that the `ComputeMgr` destructor will be called either in between consecutive `run` commands.

How about we move this code that you just added from the `colvarproxy` destructor into a regular member function, and define a function `colvarproxy::post_run()` that calls it?

This way we handle the VMD side as well without having to rely on a destructor.

Reply:
Of course, after writing I realized that calling a function called `post_run()` is not appropriate in VMD.
So it should call just the new function containing the code you added (`colvarproxy::close_files()`?). As for actually saving data from memory to the disk, this should be left to the user in an interactive code like VMD.

Reply:
> So it should call just the new function containing the code you added (colvarproxy::close_files()?). 

Yup, sounds logical.

> As for actually saving data from memory to the disk, this should be left to the user in an interactive code like VMD.

In general yes. The limiting case would be writing a colvars.traj file by iterating over a VMD trajectory: there will be no signal to flush the file because the process has no well-defined termination. But at least one way the user can control that is through `cv reset`, which calls `proxy->flush_output_streams();`.

Reply:
I did this change in 65563f02 (pushed directly to `master`). I opted for calling `flush_output_streams()` instead of `close_files()` within `post_run()`, which seemed safer. (Some files may need to stay open between `run` commands, but should definitely be closed when all objects are destroyed.)